libraries/standard/mqtt/utest/mqtt_lightweight_utest.c

```diff
@@ -0,0 +1,496 @@
+{
+    pNetworkBuffer->pBuffer = mqttBuffer;
+    pNetworkBuffer->size = MQTT_TEST_BUFFER_LENGTH;
+}
+
+/**
+ * @brief Initialize pConnectInfo using test-defined macros.
+ *
+ * @param[in] pConnectInfo MQTT CONNECT packet parameters.
+ */
+static void setupConnectInfo( MQTTConnectInfo_t * const pConnectInfo )
+{
+    pConnectInfo->cleanSession = true;
+    pConnectInfo->pClientIdentifier = CLIENT_IDENTIFIER;
+    pConnectInfo->clientIdentifierLength = CLIENT_IDENTIFIER_LEN;
+    pConnectInfo->keepAliveSeconds = 0;
+    pConnectInfo->pUserName = MQTT_TEST_USERNAME;
+    pConnectInfo->userNameLength = MQTT_TEST_USERNAME_LEN;
+    pConnectInfo->pPassword = MQTT_TEST_PASSWORD;
+    pConnectInfo->passwordLength = MQTT_TEST_PASSWORD_LEN;
+}
+
+/**
+ * @brief Initialize pWillInfo using test-defined macros.
+ *
+ * @param[in] pWillInfo Last Will and Testament.
+ */
+static void setupWillInfo( MQTTPublishInfo_t * const pWillInfo )
+{
+    pWillInfo->pPayload = NULL;
+    pWillInfo->payloadLength = 0;
+    pWillInfo->pTopicName = CLIENT_IDENTIFIER;
+    pWillInfo->topicNameLength = CLIENT_IDENTIFIER_LEN;
+    pWillInfo->dup = true;
+    pWillInfo->qos = MQTTQoS0;
+    pWillInfo->retain = true;
+}
+
+/**
+ * @brief Encode remaining length into pDestination for packet serialization
+ * using MQTT v3.1.1 spec.
+ *
+ * @param[in] pDestination Buffer to write encoded remaining length.
+ * @param[in] length Actual Remaining length.
+ */
+static size_t encodeRemainingLength( uint8_t * pDestination,
+                                     size_t length )
+{
+    uint8_t lengthByte;
+    uint8_t * pLengthEnd = NULL;
+    size_t remainingLength = length;
+
+    TEST_ASSERT_NOT_NULL( pDestination );
+
+    pLengthEnd = pDestination;
+
+    /* This algorithm is copied from the MQTT v3.1.1 spec. */
+    do
+    {
+        lengthByte = ( uint8_t ) ( remainingLength % 128U );
+        remainingLength = remainingLength / 128U;
+
+        /* Set the high bit of this byte, indicating that there's more data. */
+        if( remainingLength > 0U )
+        {
+            UINT8_SET_BIT( lengthByte, 7 );
+        }
+
+        /* Output a single encoded byte. */
+        *pLengthEnd = lengthByte;
+        pLengthEnd++;
+    } while( remainingLength > 0U );
+
+    return ( size_t ) ( pLengthEnd - pDestination );
+}
+
+/**
+ * @brief Encode UTF-8 string and its length into pDestination for
+ * packet serialization.
+ *
+ * @param[in] pDestination Buffer to write encoded remaining length.
+ * @param[in] length Actual Remaining length.
+ */
+static size_t encodeString( uint8_t * pDestination,
+                            const char * source,
+                            uint16_t sourceLength )
+{
+    uint8_t * pBuffer = NULL;
+
+    /* Typecast const char * typed source buffer to const uint8_t *.
+     * This is to use same type buffers in memcpy. */
+    const uint8_t * pSourceBuffer = ( const uint8_t * ) source;
+
+    TEST_ASSERT_NOT_NULL( pDestination );
+
```

Comment:
Why no `TEST_ASSERT_NOT_NULL( source );` here?

Reply:
Good point. I think it would be more appropriate to just call this `pBuffer` instead.